### PR TITLE
Array type checking

### PR DESCRIPTION
**Pull request by woocash2**

### CODE REVIEW

File: app/src/main/kotlin/compiler/diagnostics/Diagnostic.kt [hunk 1]
```diff
@@ -303,6 +303,20 @@ sealed interface Diagnostic {
             ) : TypeCheckingError(listOf(yieldStatement)) {
                 override val errorMessage = "The yield occurs in non-generator function."
             }
+
+            class ExtractionFromNonArrayType(
```
Comment (ignacy123) [inline app/src/main/kotlin/compiler/diagnostics/Diagnostic.kt:307]: Maybe
```suggestion
            class AccessOfNonArrayType(
```

File: app/src/test/kotlin/compiler/analysis/TypeCheckerTest.kt [hunk 2]
```diff
@@ -1259,4 +1259,245 @@ class TypeCheckerTest {
 
         assertEquals(listOf<Diagnostic>(TypeCheckingError.YieldInNonGeneratorFunction(generatorYield)), diagnosticsList)
     }
+
+    // czynność f() -> Nic {
+    //     zm x: Tablica<Liczba> = nowa tablica(5, 0)
+    // }
+
+    @Test
+    fun `test array allocation has array type`() {
```
Comment (ignacy123) [inline app/src/test/kotlin/compiler/analysis/TypeCheckerTest.kt:1268]: We definitely want a similar test for multidimensional arrays.

File: app/src/main/kotlin/compiler/analysis/TypeChecker.kt [hunk 3]
```diff
@@ -173,9 +173,30 @@ class TypeChecker(private val nameResolution: Map<Ref<AstNode>, Ref<NamedNode>>,
...
+                is Expression.ArrayLength -> {
+                    val innerExpressionType = checkExpression(expression.expression) ?: return null
+
+                    if (innerExpressionType is Type.Array)
+                        return Type.Number
+
+                    report(TypeCheckingError.LengthOfNonArrayType(expression.expression, innerExpressionType))
+                }
+                is Expression.ArrayAllocation -> {
+                    val elementType = checkExpression(expression.initialization) ?: return null
```
Comment (ignacy123) [inline app/src/main/kotlin/compiler/analysis/TypeChecker.kt:197]: Please merge the new strategy of array initialization - this bit should get a bit more complex.

File: app/src/test/kotlin/compiler/analysis/TypeCheckerTest.kt [hunk 4]
```diff
@@ -1259,4 +1259,245 @@ class TypeCheckerTest {
 
         assertEquals(listOf<Diagnostic>(TypeCheckingError.YieldInNonGeneratorFunction(generatorYield)), diagnosticsList)
     }
+
+    // czynność f() -> Nic {
+    //     zm x: Tablica<Liczba> = nowa tablica(5, 0)
```
Comment (ignacy123) [inline app/src/test/kotlin/compiler/analysis/TypeCheckerTest.kt:1264]: This syntax will need an update once we (finally) agree on the final version.

**Review by ignacy123 — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/compiler/analysis/NameResolver.kt [hunk 5]
```diff
@@ -265,9 +265,19 @@ object NameResolver {
                     analyzeNode(node.resultWhenTrue, currentScope)
                     analyzeNode(node.resultWhenFalse, currentScope)
                 }
-                is Expression.ArrayElement -> TODO()
-                is Expression.ArrayLength -> TODO()
-                is Expression.ArrayAllocation -> TODO()
+
+                is Expression.ArrayElement -> {
+                    analyzeNode(node.expression, currentScope)
+                    analyzeNode(node.index, currentScope)
```
[the remark this answers is not in the document]
Comment (woocash2) [reply app/src/main/kotlin/compiler/analysis/NameResolver.kt:271]: Good idea

File: app/src/main/kotlin/compiler/diagnostics/Diagnostic.kt (hunk 1 above)
Comment (woocash2) [reply app/src/main/kotlin/compiler/diagnostics/Diagnostic.kt:307]: AccesFrom?

Comment (ignacy123) [reply app/src/main/kotlin/compiler/diagnostics/Diagnostic.kt:307]: Sounds good :+1:

File: app/src/main/kotlin/compiler/analysis/NameResolver.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (ignacy123) [reply app/src/main/kotlin/compiler/analysis/NameResolver.kt:271]: Feel free to update it then :stuck_out_tongue:

[the remark this answers is not in the document]
Comment (woocash2) [reply app/src/main/kotlin/compiler/analysis/NameResolver.kt:271]: oh yeah, forgot

File: app/src/main/kotlin/compiler/analysis/TypeChecker.kt [hunk 6]
```diff
@@ -173,9 +173,33 @@ class TypeChecker(private val nameResolution: Map<Ref<AstNode>, Ref<NamedNode>>,
...
                 }
-                is Expression.ArrayElement -> TODO()
-                is Expression.ArrayLength -> TODO()
-                is Expression.ArrayAllocation -> TODO()
+
+                is Expression.ArrayElement -> {
+                    val innerExpressionType = checkExpression(expression.expression) ?: return null // if null then something must've gone bad in the recursion
+
+                    if (innerExpressionType is Type.Array) {
+                        checkExpression(expression.index, Type.Number)
```
Comment (ignacy123) [inline app/src/main/kotlin/compiler/analysis/TypeChecker.kt:181]: Don't we need failure handling here as well? :thinking: 
Add a test for this case too, something like `arr[prawda]`

File: app/src/test/kotlin/compiler/e2e/CorrectProgramsE2eTest.kt [hunk 7]
```diff
@@ -72,10 +72,10 @@ class CorrectProgramsE2eTest {
             czynność dopóKI_pomiń_przerwij() {
```
Comment (ignacy123) [inline app/src/test/kotlin/compiler/e2e/CorrectProgramsE2eTest.kt:72]: I know it's not your code, but we can fix this here :smile:

**Review by ignacy123 — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/compiler/analysis/TypeChecker.kt (hunk 6 above)
Comment (woocash2) [reply app/src/main/kotlin/compiler/analysis/TypeChecker.kt:181]: Yeee

Comment (woocash2) [reply app/src/main/kotlin/compiler/analysis/TypeChecker.kt:181]: What failure handling specifically? `checkExpression` in 181 already appends to diagnostics if the type mismatches.
I'll add a test.

Comment (ignacy123) [reply app/src/main/kotlin/compiler/analysis/TypeChecker.kt:181]: You're right, we don't need to add anything here

File: app/src/main/kotlin/compiler/ast/AstNode.kt [hunk 8]
```diff
@@ -28,7 +28,7 @@ sealed interface AstNode {
                     is Expression.ArrayAllocation -> {
                         var arrayValues = this.initialization.joinToString { it.toSimpleString() }
                         if (this.initializationType == Expression.ArrayAllocation.InitializationType.ONE_VALUE) arrayValues += "..."
-                        "ciąg ${this.type}[${this.size.toSimpleString()}]{ $arrayValues }"
+                        "alokacja ${this.elementType}[${this.size.toSimpleString()}]{ $arrayValues }"
```
Comment (ignacy123) [inline app/src/main/kotlin/compiler/ast/AstNode.kt:31]: I think we (finally) agreed on "ciąg" as the allocation keyword, so update this here and in code snippets in tests.

**Review by ignacy123 — CHANGES_REQUESTED** [review]

**Review by ignacy123 — APPROVED** [review]

**Pull request merged**